pkg/resources/management.cattle.io/v3/project/validator_test.go

```diff
@@ -901,7 +908,410 @@ func TestProjectValidation(t *testing.T) {
 			wantAllowed: false,
 			wantErr:     true,
 		},
+		{
```

Comment:
I think it would be worth splitting the test into two. The old test was testing mostly quota related stuff, we could introduce a new `TestProjectLimits` or something that just tests the limits. This can also shorten the amount of code written by having a test table that only takes in the `v3.ContainerResourceLimit` struct instead of the whole `v3.Project`.

Something like:

```go
func TestProjectLimit(t *testing.T) {
	t.Parallel()
	tests := []struct {
		name        string
		newLimit    *v3.ContainerResourceLimit
		oldLimit    *v3.ContainerResourceLimit
		wantAllowed bool
		wantErr     bool
	}{
		{
			name: "container default resource limit specified cpu and memory request and limits",
			newLimit: &v3.ContainerResourceLimit{
				RequestsCPU:    "1m",
				LimitsCPU:      "20m",
				RequestsMemory: "1Mi",
				LimitsMemory:   "20Mi",
			},
			wantAllowed: true,
		},
		{
			name:        "container default resource limit with no requests or limits",
			newLimit:    &v3.ContainerResourceLimit{},
			wantAllowed: true,
		},

```

Imo this would make the tests more maintainable

Reply:
Personally I'm not a fan of separating a test for a single function into two separate functions. If the function under test is updated, you now have two separate tests to debug/fix, and its harder to see what the effective coverage for a single test is since you now have two test functions, each with their own methods of mocking/setting up the tests. 

For those reasons I would vote against this, though this is mostly a style concern so ultimately it's up to max.

Reply:
I agree with Tom and, in general, prefer to split tests into separate cases based on the aspects of behavior being tested. The quota setup may get in the way of setting up tests for limits. Create/update would have multiple separate test cases, but the cover tool would still show how many statements are covered in the package as pertains to create/update.

Reply:
> I agree with Tom and, in general, prefer to split tests into separate cases based on the aspects of behavior being tested. 

We've argued about this at length before. I think we will just need to agree to disagree here.

> but the cover tool would still show how many statements are covered in the package as it pertains to create/update.

The cover tool only indicates if a line of code was run, not if it is effectively tested. Here I'm speaking about coverage of the logic, not necessarily the code, which is not detectable by the coverage tool.

Reply:
The cover tool will show coverage of the logic (expressed in statements), just not of all possible values. Keeping the tests in a single case won't help with this either.

Reply:
I can give my reasoning as to why I added the suggestion if it helps anybody. 

The current test already separates the tests into two categories: quotas and limits, but it's not made explicit. There aren't any test cases that covers both quotas & limits checks.

We might as well have something like:

```go
func TestProjectValidation(t *testing.T) {
  // Setup table for quota validation
  ...
  // Test quota
  ...
  // Setup table for limit validation
  ...
  // Test limit
  ...
}
```

and the test would be identical in the results, but each table could have specialized struct that makes it easier to review,add,modify,troubleshoot tests (IMO). **In my opinion, it's harder to find the tests that test limits to see if we have enough coverage when these tests can be randomly placed in a table full of other tests.**

We could separate the tests as mentioned above, or we could also separate into different functions. We could even test the method `validateContainerDefaultResourceLimit` at length, and instead test a few things only for the `Admit()` call.

I'm okay with whatever you choose to do though, just wanted to put my 2 cents here.